v0/destinations/one_signal/transform.js

```diff
@@ -0,0 +1,286 @@
+  }
+  if (payload.last_active) {
+    payload.last_active = toUnixTimestamp(payload.last_active);
+  }
+
+  // If playerId is present, creating Edit Device Response for Editing a devic using the playerId
+  if (playerId) {
+    endpoint = `${endpoint}/${playerId}`;
+    payload.tags = tags;
+    return responseBuilder(payload, endpoint, message.type);
+  }
+  // Creating response for creation of new device or updation of an existing device
+  populateDeviceType(message, payload);
+  const responseArray = [];
+  payload.tags = tags;
+  // Creating a device with email as asn identifier
+  if (emailDeviceType) {
+    const emailDevicePayload = { ...payload };
+    emailDevicePayload.device_type = 11;
+    emailDevicePayload.identifier = getFieldValueFromMessage(message, "email");
+    if (isDefinedAndNotNullAndNotEmpty(emailDevicePayload.identifier)) {
+      responseArray.push(
+        responseBuilder(emailDevicePayload, endpoint, message.type)
+      );
+    }
+  }
+  // Creating a device with phone as asn identifier
+  if (smsDeviceType) {
+    const smsDevicePayload = { ...payload };
+    smsDevicePayload.device_type = 14;
+    smsDevicePayload.identifier = getFieldValueFromMessage(message, "phone");
+    if (isDefinedAndNotNullAndNotEmpty(smsDevicePayload.identifier)) {
+      responseArray.push(
+        responseBuilder(smsDevicePayload, endpoint, message.type)
+      );
+    }
+  }
+  // checking if device_type is defined or not and checking 0 for device_type iOS
+  if (payload.device_type || payload.device_type === 0) {
+    responseArray.push(responseBuilder(payload, endpoint, message.type));
+  }
+  if (!responseArray.length) {
+    throw new CustomError(
+      "Correct identifier is required for creating a device (identify call)",
+      400
+    );
+  }
+  return responseArray;
+};
+
+/**
+ * This function is used to build the response for track call.
+ * It is used to edit the OneSignal tags using external_user_id.
+ * @param {*} message
+ * @param {*} param1
+ * @returns
+ */
+const trackResponseBuilder = (message, { Config }) => {
+  const { appId, eventAsTags, allowedProperties } = Config;
+  const event = get(message, "event");
+  let { endpoint } = ENDPOINTS.TRACK;
+  const externalUserId = getFieldValueFromMessage(message, "userId");
+  if (!event) {
+    throw new CustomError(
+      "[OneSignal]: event is not present in the input payloads",
+      400
+    );
+  }
+  if (!externalUserId) {
+    throw new CustomError(
+      "[OneSignal]: userId is required for track events/updating a device",
+      400
+    );
+  }
+  endpoint = `${endpoint}/${appId}/users/${externalUserId}`;
+  const payload = {};
+  const tags = {};
+  /* Populating event as true in tags.
+  eg. tags: {
+    "event_name": true
+  }
+  */
+  tags[event] = true;
+  // Populating tags using allowed properties(from dashboard)
+  const properties = get(message, "properties");
+  allowedProperties.forEach(item => {
```

Comment:
check if `allowedProperties` is defined and an array

Reply:
Do we need to add this check if it is taken from dashboard only.

Reply:
As it is coming from some other services we can't rely on them. We should check if the correct value is coming or not.

Reply:
Done